juju/juju
# leases: add raft-FSM lease store

## Description of change

Implements a new core/lease.Store as a raft FSM. Instead of running a lease manager for singular and leadership inside each state in the controller agent's state pool, we run one lease manager for the controller agent in the dependency engine. The manager now passes namespace and model uuid along with the lease name - together these are the lease key.

That manager talks to a lease store that reads data directly from the FSM, but all writes (claim, extend, expire) are sent over the pubsub hub to the raft leader to be applied there. We also run another global clock updater to increment the time in the raft lease store (using the same pubsub request mechanism) alongside the clock updater talking to mongo.

The raft lease store still updates the mongo database (the leaseholders collection, to avoid stepping on the toes of the state lease store), but only when leases are claimed or expired, which occur far less frequently than extensions (every 30 seconds).

Lease changes (claims and expiries) are also written to `/var/log/juju/lease.log`. In a multi-controller situation these will need to be combined to get a complete history of lease changes. These logs are rotated and compressed.

The two lease systems (mongo and raft) are both running - which one is in use is controlled by the `raft-leases` feature flag. This means it _can_ be switched on a running controller, although since the lease state isn't synchronised between them there's likely to be some fallout - it's possible for two units to both think they're the leader, for example. I'd advise not switching it on a controller you wouldn't be ok with throwing away for now.

## QA steps

* Bootstrap a controller with `--config="features=[raft-leases]"` and including `juju.worker.lease=TRACE` in `logging-config`.
* Enable HA.
* Deploy percona-cluster and 3 units of wordpress.
* See that one wordpress unit becomes the leader for the application (in status, in logging for `juju.worker.lease.raft` and in the mongo `leaseholders` collection).
* Remove that unit. Another wordpress unit should claim leadership when the original lease expires.
* Use `juju run -m controller <machine> juju-engine-report | less +/raft:` to determine the raft leader. Remove that machine. Once things stabilise, confirm that the lease operations are continuing in a two-controller voter/nonvoter configuration. 
* Add a new controller machine by running `enable-ha` again. The lease manager on that machine should join in with the normal lease operations.
